nsqd/topic.go

```diff
@@ -226,30 +226,36 @@ func (t *Topic) messagePump() {
 		backendChan = t.backend.ReadChan()
 	}
 
+	doUpdate := func() {
+		chans = chans[:0]
+		for _, c := range t.channelMap {
+			chans = append(chans, c)
+		}
+		if len(chans) == 0 || t.IsPaused() {
+			memoryMsgChan = nil
+			backendChan = nil
+		} else {
+			memoryMsgChan = t.memoryMsgChan
+			backendChan = t.backend.ReadChan()
+		}
+	}
+
 	for {
 		select {
+		case <-t.channelUpdateChan:
+			t.RLock()
+			doUpdate()
+			t.RUnlock()
+			continue
+
 		case msg = <-memoryMsgChan:
 		case buf = <-backendChan:
 			msg, err = decodeMessage(buf)
 			if err != nil {
 				t.ctx.nsqd.logf("ERROR: failed to decode message - %s", err)
 				continue
 			}
-		case <-t.channelUpdateChan:
```

Comment:
I think the crux of the problem was actually that the select statement here could possibly never choose this path, or more likely, choose one of the message chan paths before choosing this one.